### PR TITLE
Testing Modifications

### PR DESCRIPTION
**Pull request by pearkes**
This refactors various pieces in the approach to setting up tests for providers, in order to help make it possible for HashiCorp to run regular acceptance tests that create, verify, and destroy resources that are discovered via the library.

It makes the workflow for testing each provider look like this, with AWS as an example:

```
$ export AWS_ACCESS_KEY_ID=... AWS_ACCESS_KEY_SECRET=... AWS_REGION=...
$ cd test/tf/aws; terraform init; terraform apply
$ cd ../../; go test -v ./provider/aws
…
PASS
```

But there is one major caveat to this simplification – testing no longer occurs from within provider infrastructure. 

In many of these providers, you can implicit authenticate your requests based on where go-discover is running (eg Instance Profiles/IAM roles in AWS). With this new testing approach, go-discover is only ever executed where the tests are being run from, not utilizing that form of authorization. 

However, after looking through the provider implementations, it is clear that go-discover does little different when provided static credentials vs. using these role-based approaches. Most of the logic to choose the method of authentication is handled by the various client libraries maintained by the cloud providers and is outside of go-discover.

We have to trust the client libraries will manage authentication as appropriate. By doing that, we can simplify the testing strategy and ease the automation of running these in CI. I also updated the README, modified some tests to match the provided test Terraform configuration, and several other cosmetic changes. No changes to go-discover were made otherwise. Definitely open to feedback on this approach!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,6 +30,10 @@ function.
  * Scaleway [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/scaleway/scaleway_discover.go#L14-L22)
  * SoftLayer [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/softlayer/softlayer_discover.go#L16-L25)
 
+HashiCorp maintains acceptance tests that regularly run real resources to
```
Comment (chelseakomlo) [inline README.md]: Maybe "allocate and run tests within real resources" or something similar.

File: README.md [hunk 2]
```diff
@@ -107,4 +111,141 @@ For complete API documentation, see
...
+}
+
+# 5. Export the Credentials for the client
+export ARM_CLIENT_ID=client_id
+export ARM_CLIENT_SECRET=client_secret
+export ARM_TENANT_ID=tenant_id
+export ARM_SUBSCRIPTION_ID=subscription_id
+
+# 6. Test the credentials
+$ az vm list-sizes
```
Comment (chelseakomlo) [inline README.md]: This also needs the location `az vm list-sizes --location=<location>`

**Comment by pearkes** [comment]
@preetapan @chelseakomlo Thanks for taking a look! I believe the previous VMSS test explicitly verified that using an [Azure Role Definition](https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles) worked without static credentials. This was helpful, but after looking through how the provider [authenticates](https://github.com/hashicorp/go-discover/blob/master/provider/azure/azure_discover.go#L52-L81) to make API calls over the internet, I assumed purely static authorization would suffice for the test (see description above, same general point).

**Comment by pearkes** [comment]
Oh and to your point @chelseakomlo, tests currently log if they are skipped.

```
 $ go test -v ./provider/aws/
=== RUN   TestAddrs
--- SKIP: TestAddrs (0.00s)
	aws_discover_test.go:23: AWS credentials or region missing
PASS
ok  	github.com/hashicorp/go-discover/provider/aws	0.027s
```

File: README.md (hunk 2 above)
Comment (pearkes) [reply README.md]: Good catch!

**Pull request merged**